worldedit-core/src/main/java/com/fastasyncworldedit/core/function/mask/AngleMask.java

```diff
@@ -126,38 +128,42 @@ protected boolean testSlope(Extent extent, int x, int y, int z) {
         }
     }
 
-    private boolean adjacentAir(Extent extent, MutableBlockVector3 mutable) {
-        int x = mutable.x();
-        int y = mutable.y();
-        int z = mutable.z();
-        if (!mask.test(extent, mutable.setComponents(x + 1, y, z))) {
+    // for optimal performance, this method should be called with base being a CharFilterBlock
+    private boolean adjacentAir(Extent extent, BlockVector3 base) {
+        int y = base.y();
+        // we expect the data for blocks above and below to be loaded already
+        // in which case caching/reading from cache has more overhead
+        if (y != maxY && !fastMask.test(base.getStateRelativeY(extent, 1))) {
             return true;
         }
-        if (!mask.test(extent, mutable.setComponents(x - 1, y, z))) {
+        if (y != minY && !fastMask.test(base.getStateRelativeY(extent, -1))) {
             return true;
         }
-        if (!mask.test(extent, mutable.setComponents(x, y, z + 1))) {
+        // other positions might be in different chunks, go a slower, cached path there
```

Comment:
Potentially worth coming back to this after the merge of this and #3051 - if the cache performance is still poor then probably worth testing if x/z are actually on the edges of the chunk. Maybe this could be implemented into the cache mask constructor - we can potentially look at ways of compressing the data further.